gensim/summarization/bm25.py

```diff
@@ -116,6 +119,13 @@ def _initialize(self, corpus):
                 negative_idfs.append(word)
         self.average_idf = float(idf_sum) / len(self.idf)
 
+        if self.average_idf < 0:
```

Comment:
What does IDF == 0 mean, is that edge case OK?

Reply:
Average IDF < 0 iff the product of (N - n_w) / n_w is less than one, where N is the corpus size and n_w is the number of documents containing word w. Due to the Zipf's empirical law, this should not happen for large enough corpora (few words are very common, most words occur rarely).

Words w with IDF_w < 0 will contribute negatively to the BM25 score, i.e. documents with more words w in common with the query will receive smaller BM25 score. This seems “wrong”, but it is actually what we want when we are dealing with stopwords.

Reply:
> Average IDF < 0 iff the product of (N - n_w) / n_w is less than one

You mean average over all such products = averaged over all `_w`?

Anyway, my question was around the edge case `== 0` (as opposed to `< 0`). It's probably not important, just wondering whether we can capture more "weird malformed inputs" by extending the inequality.

Reply:
> > Average IDF < 0 iff the product of (N - n_w) / n_w is less than one
> 
> You mean average over all such products = averaged over all `_w`?

Average_w IDF_w < 0
iff Average_w log (N - n_w) / n_w < 0
iff Product_w (N - n_w) / n_w < 1.

> Anyway, my question was around the edge case == 0 (as opposed to < 0). It's probably not
> important, just wondering whether we can capture more "weird malformed inputs" by extending
> the inequality. "Zeros" are often interesting edge cases (empty inputs, empty dictionaries etc).

IDF_w = 0 for words w that occur precisely in one half of the corpus. There are a lot of such words in the example from the original post:
```python
>>> bm25.average_idf
-0.2514746738178282
>>> bm25.idf
{'A': 0.0, 'constellation': 0.0, 'is': 0.0, 'a': 0.0, 'group': 0.0, 'of': -0.06286866845445704, 'stars': 0.0, 'that': 0.0, 'are': -0.06286866845445704, 'considered': 0.0, 'to': 0.0, 'form': 0.0, 'imaginary': 0.0, 'outlines': 0.0, 'or': 0.0, 'meaningful': 0.0, 'patterns': 0.0, 'on': 0.0, 'the': -0.06286866845445704, 'celestial': -0.06286866845445704, 'sphere.': -0.06286866845445704, 'The': 0.0, '88': 0.0, 'modern': 0.0, 'constellations': 0.0, 'formally': 0.0, 'defined': 0.0, 'regions': 0.0, 'sky': 0.0, 'together': 0.0, 'covering': 0.0, 'entire': 0.0}
```
You may notice that the mean of `bm25.idf` is not `bm25.average_idf`. That's because BM25 floors IDF_w by setting all negative IDF_w to EPSILON times average IDF, where EPSILON is 0.25.

Reply:
On a related note, some of us would appreciate support for inline math in GitHub comments.

Reply:
Great, thank you for the detailed analysis @Witiko